spec/p2p/v0.34/addressbook.md

```diff
@@ -0,0 +1,392 @@
+thorough this document, but it can be summarized as follows.
+A new peer has its address stored in a bucket of new address, as the node does
+not have much information about it.
+When a peer is marked as a [good peer](#good-peers) by some reactor (protocol),
+its address is moved (upgraded) to a bucket of old addresses.
+Old addresses are therefore of peers about which the node has good information.
+When a bucket of old addresses becomes full, one of its addresses is moved
+(downgraded) to a bucket of new addresses.
+When a bucket of new addresses becomes full, one of its addresses is dropped.
+New addresses, therefore, are deemed less important than old addresses, being
+dropped easily by the address book.
+
+The address book can keep multiple network addresses for the same node ID, but
+this number should be limited.
+A new address with a node ID that is already present in the address book is
+**not added** when:
+
+- The last address added with the same node ID is stored in an old bucket, so
+  it is considered a "good" address
+- There are addresses associated to the same node ID stored in
+  `maxNewBucketsPerAddress = 4` distinct buckets
+- Randomly, with a probability that increases exponentially with the number of
+  buckets in which there is an address with the same node ID.
+  So, a new address for a node ID which is already present in one bucket is
+  added with 50% of probability; if the node ID is present in two buckets, the
+  probability decreases to 25%; and if it is present in three buckets, the
+  probability is 12.5%.
+
+The bucket to which a new address is added is computed based on the address
+itself and its source address.
+An address is not added to a bucket if it is already present there.
+This should not happen with addresses with new node IDs, but it can happen with
+addresses with known node IDs.
+In fact, as the mapping of buckets for new addresses is mainly determined by
+the network group of the source address, the same address can be present in
+more than one bucket provided that it was added with different source addresses.
+
+The new address is also added to the `addrLookup` table, which stores
+`knownAddress` entries indexed by their node IDs.
+If the new address is from an unknown peer, a new entry is added to the
+`addrLookup` table; otherwise, the existing entry is updated with the new
+address.
+Entries of this table contain, among other fields, the list of buckets where
+addresses of a peer are stored.
+The `addrLookup` table is used by most of the address book methods (e.g.,
+`HasAddress`, `IsGood`, `MarkGood`, `MarkAttempt`), as it provides fast access
+to addresses.
+
+> TODO: reorganize this description having the goals of this organization
+> explained before their actual implementation.
+
+### Errors
+
+- if the added address or the associated source address are nil
+- if the added address is invalid
+- if the added address is the local node's address
+- if the added address ID is of a banned peer
```

Comment:
"banned" is not defined at this point. Please add a forward reference to the "bad peers" section